### ISSUE TITLE
Reduce memory usage

### ISSUE THREAD

**Issue by splitice** [resin-io/resin-supervisor#690]
The supervisor node process starts at 128mb and grows up to around 200mb on our environment.

Environment:
 - Single Container
 - 32bit ARM

On 512MB devices this a significant memory cost for what is a relatively simple application.

We have been investigating methods to reduce this, patching the supervisor container to test various options. So far we have achieved a 60MB memory reduction with the following options:

```
--max_old_space_size=40 --optimize_for_size --always_compact --stack_size=700
```

This will slightly increase the CPU usage (more GC cycles, and a reduction in inlining) however we have not noticed any real performance degradation.

**Comment by CameronDiver** [comment]
Hey @splitice, firstly thanks for bringing this to our attention. I raised this for an internal discussion, and it's definitely something that we have explored in the past.

The outcome of this discussion was that we're very happy with `--optimize_for_size` and `--always_compact`, but we've experienced instabilities with the other options in the past, and we would need further information on the gains that this could provide for these options.

We would happily take patches for `--optimize_for_size` and `--always_compact` right now, if you are interested, could you open a PR? 

We would also happily accept patches for the other flags, if you find that they provide additional improvements over the first two flags and prove to be stable.

**Comment by splitice** [comment]
The stack size decrease saves only ~90KB per thread. So it's pretty minor, less than 1MB in savings total.

The max old size is actually very effective, as it provides a watermark to the GC to indicate that it needs to run. As long as persistent memory requirements do not exceed this limit all is good, 40MB seems fine. 30MB seems to run stably (at a glance, not long term) too, but a higher than current requirements value would probably be wise to prevent any crashes.

We have been running these arguments on one device as a test for a couple days without any issues (yet). I'm certain that the numbers could be smaller, I was conservative in selecting them, but the only way to know for sure is wider testing (I'm guessing multi-container supervisors might need slightly more memory?).

**Comment by splitice** [comment]
Pull request opened with the first candidate.

I increased the max old size by a further 50MB in case you think that's safer for testing.

**Comment by CameronDiver** [comment]
@splitice applying them on a testing device would be great. Also what device type have you been testing these on already?

I checked out your branch and did a little bit of testing, and it seems fairly stable, although I haven't had the time to do anything too insightful just yet.

**Comment by splitice** [comment]
We are testing on a Nano PI Neo AIR (closest equivalent is the Orange PI which is the same SoC). Any 32bit ARM device should use approximately the same amount of memory. 64bit devices may use a little bit more memory due to increase pointer sizes. 

Our testing method has just been to run with it on our development devices and keep an eye out for any crashes (one observed at 35mb, none so far at 50mb).

**Comment by Page-** [comment]
In relation to reducing memory usage I believe there could also be an improvement from nodejs 8 as I believe it includes v8 changes to reduce memory usage on low memory devices.  I opened an issue about that on https://github.com/resin-io/resin-supervisor-base/issues/8 but it might be something you find useful to try

**Comment by splitice** [comment]
@Page- There are other areas improvements can be had for sure. I was not aware of improvements in nodejs 8 but it does not suprise me.

We recently optimized the NodeJS side of our platform and got 80MB applications down to approx 25MB the actions we took were:
1. Command Line options similar to this PR
2. Reducing large / bloated external dependencies, in particular removal of ```request```, ```mime``` (all dependencies using this patched to use ```mime-types``` instead for single source of mimes) and ```send```. This included the creation of a low memory fork of express (https://github.com/Halleyassist/express)

We are already running on NodeJS 9 so no gains for us there.
